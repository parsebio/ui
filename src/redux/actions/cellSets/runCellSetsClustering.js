import {
  CELL_SETS_ERROR, CELL_SETS_CLUSTERING_UPDATING,
} from 'redux/actionTypes/cellSets';

import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';
import fetchWork from 'utils/work/fetchWork';
import getCellSets from 'redux/selectors/cellSets/getCellSets';

const runCellSetsClustering = (experimentId, resolution) => async (dispatch, getState) => {
  const {
    error, updatingClustering, loading,
  } = getCellSets()(getState().cellSets);

  const { experimentSettings: { processing } } = getState();

  const { method } = processing.configureEmbedding.clusteringSettings;

  if ((loading && updatingClustering) || error) {
    return null;
  }

  const body = {
    name: 'ClusterCells',
    cellSetName: 'Louvain clusters',
    type: method,
    cellSetKey: 'louvain',
    config: {
      resolution,
    },
  };

  dispatch({
    type: CELL_SETS_CLUSTERING_UPDATING,
  });

  const timeout = getTimeoutForWorkerTask(getState(), 'ClusterCells');

  try {
    await fetchWork(
      experimentId,
      body,
      getState,
      dispatch,
      {
        timeout,
        broadcast: true,
      },
    );
  } catch (e) {
    dispatch({
      type: CELL_SETS_ERROR,
      payload: {
        experimentId,
        error: e,
      },
    });
  }
};

export default runCellSetsClustering;
