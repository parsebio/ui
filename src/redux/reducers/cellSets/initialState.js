const initialState = {
  properties: {},
  hierarchy: [],
  initialLoadPending: true,
  loading: false,
  error: false,
  updatingClustering: false,
  hidden: new Set(),
};

export default initialState;
