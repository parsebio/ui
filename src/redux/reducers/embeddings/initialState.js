/**
 * The store has the following format:
 * {
 *    embeddingType: {
 *      data: {xValues:[1,2], yValues:[5,3], cellIds:[0,3]},
 *      loading: false,
 *      error: false
 *    }
 * }
 */

const initialEmbeddingState = {
  data: {},
  loading: true,
  error: false,
};

const initialState = {};

export { initialEmbeddingState };
export default initialState;
