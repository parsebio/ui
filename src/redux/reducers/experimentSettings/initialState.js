const metaInitialState = {
  loading: false,
  completingStepError: false,
  loadingSettingsError: false,
  changedQCFilters: new Set(),
};

const initialState = {
  info: {
    experimentId: null,
    experimentName: null,
    sampleIds: [],
    pipelineVersion: null,
  },
  processing: {
    meta: metaInitialState,
  },
  originalProcessing: {},
};

export { metaInitialState };
export default initialState;
