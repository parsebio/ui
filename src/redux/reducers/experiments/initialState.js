const experimentTemplate = {
  name: null,
  description: 'Analysis description',
  id: null,
  createdAt: null,
  updatedAt: null,
  notifyByEmail: true,
  sampleIds: [],
  metadataKeys: [],
  pipelineVersion: 1,
  parentExperimentId: null,
  cellLevelMetadata: null,
  isSubsetted: false,
};

const initialState = {
  meta: {
    loading: false,
    error: false,
    saving: false,
    activeExperimentId: null,
  },
  ids: [],
};

const METADATA_DEFAULT_VALUE = 'N.A.';

export default initialState;
export { experimentTemplate, METADATA_DEFAULT_VALUE };
