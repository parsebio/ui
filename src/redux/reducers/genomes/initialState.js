const initialState = {
  loading: false,
  error: false,
  public: {},
  custom: {},
};

export default initialState;
