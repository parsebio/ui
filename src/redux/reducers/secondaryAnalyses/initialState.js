const secondaryAnalysisTemplate = {
  name: null,
  id: null,
  createdAt: null,
  kit: null,
  chemistryVersion: null,
  numOfSamples: null,
  numOfSublibraries: null,
  files: { data: {} },
};

const initialState = {
  meta: {
    loading: false,
    error: false,
    saving: false,
    activeSecondaryAnalysisId: null,
  },
  ids: [],
};

export { secondaryAnalysisTemplate };
export default initialState;
