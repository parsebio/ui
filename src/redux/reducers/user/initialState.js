const initialState = {
  current: null,
  isAuthenticated: false,
};

export default initialState;
