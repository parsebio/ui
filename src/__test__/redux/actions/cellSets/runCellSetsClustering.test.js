import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import runCellSetsClustering from 'redux/actions/cellSets/runCellSetsClustering';

import initialState from 'redux/reducers/cellSets/initialState';
import fetchWork from 'utils/work/fetchWork';

enableFetchMocks();
const mockStore = configureStore([thunk]);

jest.mock('utils/work/fetchWork');

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

const startDate = '2021-01-01T00:00:00';

describe('runCellSetsClustering action', () => {
  const experimentId = '1234';

  const backendStatus = { [experimentId]: { status: { pipeline: { startDate } } } };
  const experimentSettingsStore = {
    processing: { configureEmbedding: { clusteringSettings: { method: 'louvain' } } },
  };

  beforeEach(async () => {
    jest.clearAllMocks();
    const response = new Response(JSON.stringify({}));

    fetchMock.resetMocks();
    fetchMock.mockResolvedValueOnce(response);

    fetchWork.mockImplementation(() => Promise.resolve());
  });

  it('Does not dispatch on loading state if clustering is already recomputing', async () => {
    const store = mockStore({
      cellSets: { loading: true, error: false, updatingClustering: true },
      experimentSettings: experimentSettingsStore,
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
    });

    store.dispatch(runCellSetsClustering(experimentId));
  });

  it('Does dispatch on loading state if clustering is not recomputing', async () => {
    const store = mockStore({
      cellSets: { loading: true, error: false, updatingClustering: false },
      experimentSettings: experimentSettingsStore,
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
    });

    await store.dispatch(runCellSetsClustering(experimentId));
    expect(fetchWork).toHaveBeenCalledTimes(1);
  });

  it('Does not dispatch on error state', async () => {
    const store = mockStore({
      cellSets: { loading: false, error: true },
      experimentSettings: experimentSettingsStore,
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
    });
    await store.dispatch(runCellSetsClustering(experimentId));
    expect(fetchWork).not.toHaveBeenCalled();
  });

  it('Dispatches all required actions to update cell sets clustering.', async () => {
    const store = mockStore({
      cellSets: {
        ...initialState,
        loading: false,
        error: false,
        hierarchy: [{ children: [], key: 'scratchpad' }],
        properties: {
          scratchpad: {
            cellIds: new Set(),
            color: undefined,
            name: 'Scratchpad',
            rootNode: true,
            type: 'cellSets',
          },
        },
      },
      experimentSettings: experimentSettingsStore,
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
    });

    await store.dispatch(runCellSetsClustering(experimentId, 0.5));

    expect(fetchWork).toHaveBeenCalledTimes(1);
    expect(fetchWork.mock.calls).toMatchSnapshot();
  });

  it('Dispatches error action when fetchWork fails', async () => {
    const store = mockStore({
      cellSets: { ...initialState, loading: false, error: false },
      experimentSettings: experimentSettingsStore,
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
    });

    fetchWork.mockImplementation(() => Promise.reject());

    await store.dispatch(runCellSetsClustering(experimentId, 0.5));

    expect(fetchWork).toHaveBeenCalledTimes(1);
    expect(fetchWork.mock.calls).toMatchSnapshot();
  });
});
