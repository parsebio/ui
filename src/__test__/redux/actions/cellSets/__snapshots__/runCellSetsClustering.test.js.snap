// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runCellSetsClustering action Dispatches all required actions to update cell sets clustering. 1`] = `
[
  [
    "1234",
    {
      "cellSetKey": "louvain",
      "cellSetName": "Louvain clusters",
      "config": {
        "resolution": 0.5,
      },
      "name": "ClusterCells",
      "type": "louvain",
    },
    [Function],
    [Function],
    {
      "broadcast": true,
      "timeout": 60,
    },
  ],
]
`;

exports[`runCellSetsClustering action Dispatches error action when fetchWork fails 1`] = `
[
  [
    "1234",
    {
      "cellSetKey": "louvain",
      "cellSetName": "Louvain clusters",
      "config": {
        "resolution": 0.5,
      },
      "name": "ClusterCells",
      "type": "louvain",
    },
    [Function],
    [Function],
    {
      "broadcast": true,
      "timeout": 60,
    },
  ],
]
`;
